Tasks/SonarQubePostTest/SonarQubeMetrics.ps1

```diff
@@ -0,0 +1,168 @@
+#region Public 
+
+#
+# Gets the chached quality gate status or queries the server to get the result of the quality gate if it is not cached
+# Remark: this only works for SQ version 5.3+
+#
+function GetOrFetchQualityGateStatus
+{
+    $qualityGateStatus = GetTaskContextVariable "MSBuild.SonarQube.QualityGateStatus"
+    
+    if ([String]::IsNullOrEmpty($qualityGateStatus))
+    {
+        $qualityGateStatus = FetchQualityGateStatus  
+        SetTaskContextVariable "MSBuild.SonarQube.QualityGateStatus" $qualityGateStatus              
+    }
+    
+    Assert (![String]::IsNullOrEmpty($qualityGateStatus)) "Could not fetch the quality gate status"
+    
+    return $qualityGateStatus
+}
+
+#endregion
+
+#region Private
+
+function FetchQualityGateStatus
+{
+    $analysisId = GetTaskContextVariable "MSBuild.SonarQube.AnalysisId"       
+    
+    if ([String]::IsNullOrEmpty($analysisId))
```

Comment:
This is where the blocking operation now occurs. It the code would be easier to follow if the _WaitForAnalysisToFinish_ was made at the same place as before, and this method just asserted if it hasn't been set.
The new code structure doesn't seem to offer any particular benefits in terms of ease or maintainability; it just makes it harder to work out when/where the blocking call occurs.